File: app/src/test/java/com/example/mike/mp3player/client/activities/SplashScreenEntryActivityTest.java
package com.example.mike.mp3player.client.activities;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;

import androidx.test.core.app.ActivityScenario;
import androidx.test.platform.app.InstrumentationRegistry;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

import static android.Manifest.permission.WRITE_EXTERNAL_STORAGE;
import static com.example.mike.mp3player.client.activities.SplashScreenEntryActivity.APP_TERMINATED;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class SplashScreenEntryActivityTest {

    private ActivityScenario<SplashScreenEntryActivity> scenario;

    @Before
    public void setup() {
        this.scenario = ActivityScenario.launch(SplashScreenEntryActivity.class);
    }

    @Test
    public void testNonRootTask() {
        scenario.onActivity((SplashScreenEntryActivity splashActivity) -> {
            SplashScreenEntryActivity spiedActivity = spy(splashActivity);
            Context context = InstrumentationRegistry.getInstrumentation().getContext();
            Intent intent = new Intent(context, SplashScreenEntryActivity.class);
            intent.addCategory(Intent.CATEGORY_LAUNCHER);
            intent.setAction(Intent.ACTION_MAIN);
            when(spiedActivity.getIntent()).thenReturn(intent);
            when(spiedActivity.isTaskRoot()).thenReturn(false);
            spiedActivity.onCreate(null);
            assertTrue(splashActivity.isFinishing());
        });


    }

    @Test
    public void testOnRequestPermissionResult() {
        assertTrue(true);
    }

    @Test
    public void testOnConnected() {
        scenario.onActivity((SplashScreenEntryActivity splashActivity) -> {
            SplashScreenEntryActivity spiedActivity = spy(splashActivity);
            spiedActivity.setPermissionGranted(true);
            spiedActivity.setSplashScreenFinishedDisplaying(true);
            spiedActivity.onConnected();
            verify(spiedActivity, times(1)).startActivityForResult(any(Intent.class), eq(APP_TERMINATED));
        });
    }

    @Test
    public void testOnPermissionGranted() {
        scenario.onActivity((SplashScreenEntryActivity splashActivity) -> {
            SplashScreenEntryActivity spiedActivity = spy(splashActivity);
            doNothing().when(spiedActivity).runOnUiThread(any(Runnable.class));
            spiedActivity.onPermissionGranted();
            assertTrue(spiedActivity.isPermissionGranted());
        });
    }

    @Test
    public void testOnActivityResultValidAppTerminated() {
        scenario.onActivity((SplashScreenEntryActivity splashActivity) -> {
            SplashScreenEntryActivity spiedActivity = spy(splashActivity);
            spiedActivity.onActivityResult(APP_TERMINATED, APP_TERMINATED, null);
            verify(spiedActivity, times(1)).finish();
        });
    }

    @Test
    public void testOnActivityResultInvalidAppTerminated() {
        scenario.onActivity((SplashScreenEntryActivity splashActivity) -> {
            final int NOT_APP_TERMINATED = -1;
            SplashScreenEntryActivity spiedActivity = spy(splashActivity);
            spiedActivity.onActivityResult(NOT_APP_TERMINATED, NOT_APP_TERMINATED, null);
            verify(spiedActivity, never()).finish();
        });
    }

    @Test
    public void testOnRequestPermissionsResultAccepted() {
        scenario.onActivity((SplashScreenEntryActivity splashActivity) -> {
            splashActivity = spy(SplashScreenEntryActivity.class);
            doNothing().when(splashActivity).onPermissionGranted();
            final int requestCode = 200;
            String[] permissions = new String[]{WRITE_EXTERNAL_STORAGE};
            int[] grantResults = new int[]{PackageManager.PERMISSION_GRANTED};
            splashActivity.onRequestPermissionsResult(requestCode, permissions, grantResults);
            verify(splashActivity, times(1)).onPermissionGranted();
        });
    }

    @Test
    public void testOnRequestPermissionsResultRejected() {
        scenario.onActivity((SplashScreenEntryActivity splashActivity) -> {
            final int requestCode = 200;
            String[] permissions = new String[]{WRITE_EXTERNAL_STORAGE};
            int[] grantResults = new int[]{PackageManager.PERMISSION_DENIED};
            splashActivity.onRequestPermissionsResult(requestCode, permissions, grantResults);
            assertTrue(splashActivity.isFinishing());
        });
    }

    @Test
    public void testOnRequestPermissionsResultEmpty() {
        scenario.onActivity((SplashScreenEntryActivity splashActivity) -> {
            final int requestCode = 200;
            String[] permissions = new String[]{};
            int[] grantResults = new int[]{};
            splashActivity.onRequestPermissionsResult(requestCode, permissions, grantResults);
            assertTrue(splashActivity.isFinishing());
        });
    }

//    @After
//    public void tearDown() {
//        if (null != splashScreenEntryActivity) {
//            Thread thread = splashScreenEntryActivity.getThread();
//            if (null != thread) {
//                thread.stop();
//            }
//        }
//    }

}

File: app/src/test/java/com/github/goldy1992/mp3player/client/activities/SplashScreenEntryActivityTest.kt
package com.github.goldy1992.mp3player.client.activities

import android.Manifest.permission
import android.content.Intent
import android.content.pm.PackageManager
import androidx.test.core.app.ActivityScenario
import androidx.test.platform.app.InstrumentationRegistry
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers
import org.mockito.Mockito
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class SplashScreenEntryActivityTest {
    private var scenario: ActivityScenario<SplashScreenEntryActivity>? = null
    @Before
    fun setup() {
        scenario = ActivityScenario.launch(SplashScreenEntryActivity::class.java)
    }

    @Test
    fun testNonRootTask() {
        scenario!!.onActivity { splashActivity: SplashScreenEntryActivity ->
            val spiedActivity = Mockito.spy(splashActivity)
            val context = InstrumentationRegistry.getInstrumentation().context
            val intent = Intent(context, SplashScreenEntryActivity::class.java)
            intent.addCategory(Intent.CATEGORY_LAUNCHER)
            intent.action = Intent.ACTION_MAIN
            Mockito.`when`(spiedActivity.intent).thenReturn(intent)
            Mockito.`when`(spiedActivity.isTaskRoot).thenReturn(false)
            spiedActivity.onCreate(null)
            Assert.assertTrue(splashActivity.isFinishing)
        }
    }

    @Test
    fun testOnRequestPermissionResult() {
        Assert.assertTrue(true)
    }

    @Test
    fun testOnPermissionGranted() {
        scenario!!.onActivity { splashActivity: SplashScreenEntryActivity ->
            val spiedActivity = Mockito.spy(splashActivity)
            Mockito.doNothing().`when`(spiedActivity).runOnUiThread(ArgumentMatchers.any(Runnable::class.java))
            spiedActivity.onPermissionGranted()
            Assert.assertTrue(spiedActivity.isPermissionGranted)
        }
    }

    @Test
    fun testOnActivityResultValidAppTerminated() {
        scenario!!.onActivity { splashActivity: SplashScreenEntryActivity ->
            val spiedActivity = Mockito.spy(splashActivity)
            spiedActivity.onActivityResult(SplashScreenEntryActivity.APP_TERMINATED, SplashScreenEntryActivity.APP_TERMINATED, null)
            Mockito.verify(spiedActivity, Mockito.times(1)).finish()
        }
    }

    @Test
    fun testOnActivityResultInvalidAppTerminated() {
        scenario!!.onActivity { splashActivity: SplashScreenEntryActivity ->
            val NOT_APP_TERMINATED = -1
            val spiedActivity = Mockito.spy(splashActivity)
            spiedActivity.onActivityResult(NOT_APP_TERMINATED, NOT_APP_TERMINATED, null)
            Mockito.verify(spiedActivity, Mockito.never()).finish()
        }
    }

    @Test
    fun testOnRequestPermissionsResultAccepted() {
        scenario!!.onActivity { splashActivity: SplashScreenEntryActivity ->
            Mockito.doNothing().`when`(splashActivity).onPermissionGranted()
            val requestCode = 200
            val permissions = arrayOf(permission.WRITE_EXTERNAL_STORAGE)
            val grantResults = intArrayOf(PackageManager.PERMISSION_GRANTED)
            splashActivity.onRequestPermissionsResult(requestCode, permissions, grantResults)
            Mockito.verify(splashActivity, Mockito.times(1)).onPermissionGranted()
        }
    }

    @Test
    fun testOnRequestPermissionsResultRejected() {
        scenario!!.onActivity { splashActivity: SplashScreenEntryActivity ->
            val requestCode = 200
            val permissions = arrayOf(permission.WRITE_EXTERNAL_STORAGE)
            val grantResults = intArrayOf(PackageManager.PERMISSION_DENIED)
            splashActivity.onRequestPermissionsResult(requestCode, permissions, grantResults)
            Assert.assertTrue(splashActivity.isFinishing)
        }
    }

    @Test
    fun testOnRequestPermissionsResultEmpty() {
        scenario!!.onActivity { splashActivity: SplashScreenEntryActivity ->
            val requestCode = 200
            val permissions = arrayOf<String>()
            val grantResults = intArrayOf()
            splashActivity.onRequestPermissionsResult(requestCode, permissions, grantResults)
            Assert.assertTrue(splashActivity.isFinishing)
        }
    } //    @After
//    public void tearDown() {
//        if (null != splashScreenEntryActivity) {
//            Thread thread = splashScreenEntryActivity.getThread();
//            if (null != thread) {
//                thread.stop();
//            }
//        }
//    }
}